base/errorshow.jl

```diff
@@ -29,6 +29,80 @@ ERROR: MyException: test exception
 """
 showerror(io::IO, ex) = show(io, ex)
 
+"""
+    register_error_hint(handler, exceptiontype)
+
+Register a "hinting" function `handler(io, exception)` that can
+suggest potential ways for users to circumvent errors.  `handler`
+should examine `exception` to see whether the conditions appropriate
+for a hint are met, and then generate output to `io`.
+Packages should call `register_error_hint` from within their
+[`__init__`](@ref) function.
+
+For specific exception types, `handler` is required to accept additional arguments:
+
+- `MethodError`: provide `handler(io, exc::MethodError, argtypes, kwargs)`,
+  which splits the combined arguments into positional and keyword arguments.
+
+When issuing a hint, the output should typically start with `\n`.
+
+If you define custom exception types, your `showerror` method can
+support hints by calling [`show_error_hints`](@ref).
+
+# Example
+
+```jldoctest
+julia> module Hinter
+
+       only_int(x::Int)      = 1
+       any_number(x::Number) = 2
+
+       function __init__()
+           register_error_hint(MethodError) do io, exc, argtypes, kwargs
+               if exc.f == only_int
+                   if get(io, :color, false)
+                       print(io, "\nDid you mean to call ")
+                       printstyled(io, "`any_number`?", color=:light_magenta)
+                   else
+                       print(io, "\nDid you mean to call `any_number`?")
+                   end
```

Comment:
Do you need explicit `get(io, :color, false)`?  Why not just

```suggestion
                   print(io, "\nDid you mean to call ")
                   printstyled(io, "`any_number`?", color=:light_magenta)
```

? `printstyled` is `IOContext`-aware, right?

```julia
julia> sprint() do io
           printstyled(io, "hello"; color=:red)
       end
"hello"

julia> sprint(context=stdout) do io
           printstyled(io, "hello"; color=:red)
       end
"\e[31mhello\e[39m"
```